google/resource_dataflow_job.go

```diff
@@ -0,0 +1,210 @@
+			},
+
+			"parameters": {
+				Type:     schema.TypeMap,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"on_delete": &schema.Schema{
+				Type:         schema.TypeString,
+				ValidateFunc: validation.StringInSlice([]string{"cancel", "drain"}, false),
+				Optional:     true,
+				Default:      "drain",
+				ForceNew:     true,
+			},
+
+			"project": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"state": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceDataflowJobCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	jobName := d.Get("name").(string)
+	gcsPath := d.Get("template_gcs_path").(string)
+	tempLocation := d.Get("temp_gcs_location").(string)
+	zone := d.Get("zone").(string)
+	maxWorkers := d.Get("max_workers").(int)
+	params := expandDataflowParamsStringMap(d.Get("parameters").(map[string]interface{}))
+
+	env := dataflow.RuntimeEnvironment{
+		TempLocation: tempLocation,
+		Zone:         zone,
+		MaxWorkers:   int64(maxWorkers),
+	}
+
+	request := dataflow.CreateJobFromTemplateRequest{
+		JobName:     jobName,
+		GcsPath:     gcsPath,
+		Parameters:  params,
+		Environment: &env,
+	}
+
+	job, err := config.clientDataflow.Projects.Templates.Create(project, &request).Do()
+	if err != nil {
+		return err
+	}
+
+	d.SetId(job.Id)
+
+	return resourceDataflowJobRead(d, meta)
+}
+
+func resourceDataflowJobRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	id := d.Id()
+
+	job, err := config.clientDataflow.Projects.Jobs.Get(project, id).Do()
+	if err != nil {
+		return handleNotFoundError(err, d, fmt.Sprintf("Dataflow job %s", id))
+	}
+
+	if dataflowTerminalStatesMap[job.CurrentState] {
+		log.Printf("[DEBUG] Removing resource '%s' because it is in state %s.\n", job.Name, job.CurrentState)
+		d.SetId("")
+		return nil
+	} else {
+		d.Set("state", job.CurrentState)
+	}
+
+	return nil
+}
+
+func resourceDataflowJobDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	id := d.Id()
+	requestedState, err := mapOnDelete(d.Get("on_delete").(string))
+	if err != nil {
+		return err
+	}
+	for ; !dataflowTerminalStatesMap[d.Get("state").(string)]; resourceDataflowJobRead(d, meta) {
+		job := &dataflow.Job{
+			RequestedState: requestedState,
+		}
+
+		_, err = config.clientDataflow.Projects.Jobs.Update(project, id, job).Do()
+		if gerr, ok := err.(*googleapi.Error); !ok {
+			// If we have an error and it's not a google-specific error, we should go ahead and return.
+			return err
+		} else if ok && strings.Contains(gerr.Message, "not yet ready for canceling") {
+			time.Sleep(5 * time.Second)
```

Comment:
So usually for something that requires checking a bunch of times if something is in a certain state, we would use a StateChangeConf (see all the _operation.go files for examples), which handles things like timeouts and backoff and whatnot. Is there a way to know just based on read whether the job is ready for canceling, or is it something you can only know by trying it and seeing if it works?

Reply:
You can only know by trying it, it seems to me.